File: app/src/main/java/eu/geekhome/rest/configurable/ConfigurableDtoMapper.java
package eu.geekhome.rest.configurable;

import com.geekhome.common.configurable.Configurable;
import eu.geekhome.rest.MappingException;

import javax.inject.Inject;
import java.util.List;
import java.util.stream.Collectors;

public class ConfigurableDtoMapper {

    @Inject
    private FieldDefinitionDtoMapper _fieldDefinitionDtoMapper;

    public ConfigurableDto map(Configurable configurable, List<Configurable> allConfigurables) throws MappingException {
        List<FieldDto> fields = configurable.getFieldDefinitions() == null ? null : configurable
                .getFieldDefinitions()
                .stream()
                .map(_fieldDefinitionDtoMapper::map)
                .collect(Collectors.toList());


        List<ConfigurableDto> descendants = allConfigurables
                .stream()
                .filter((x) -> x.getParent() != null && x.getParent().equals(configurable.getClass()))
                .map((x) -> this.map(x, allConfigurables))
                .collect(Collectors.toList());

        return new ConfigurableDto(configurable.getTitleRes(),
                configurable.getClass().getSimpleName(),
                fields,
                configurable.getAddNewRes(),
                configurable.getIconName(),
                descendants);
    }
}

File: app-geekhome/src/main/java/eu/geekhome/rest/configurables/ConfigurableDtoMapper.kt
package eu.geekhome.rest.configurables

import eu.geekhome.data.configurables.ConfigurableDto
import eu.geekhome.data.configurables.ConfigurableType
import eu.geekhome.data.fields.FieldDefinitionDto
import eu.geekhome.rest.MappingException
import eu.geekhome.rest.fields.FieldDefinitionDtoMapper
import eu.geekhome.domain.configurable.*
import eu.geekhome.data.localization.Resource
import java.util.stream.Collectors
import javax.inject.Inject

class ConfigurableDtoMapper @Inject constructor(
    private val fieldDefinitionDtoMapper: FieldDefinitionDtoMapper
) {
    @Throws(MappingException::class)
    fun map(configurable: Configurable): ConfigurableDto {
        var fields: List<FieldDefinitionDto>? = null
        var addNewRes: Resource? = null
        var editRes: Resource? = null
        if (configurable is ConfigurableWithFields) {
            fields = configurable
                .fieldDefinitions
                .values
                .stream()
                .map { field: FieldDefinition<*> -> fieldDefinitionDtoMapper.map(field) }
                .collect(Collectors.toList())
            addNewRes = configurable.addNewRes
            editRes = configurable.editRes
        }

        return ConfigurableDto(
            configurable.titleRes,
            configurable.descriptionRes,
            configurable.javaClass.name,
            if (configurable.parent != null) configurable.parent!!.name else null,
            getConfigurableType(configurable),
            fields,
            addNewRes,
            editRes,
            configurable.iconRaw,
            configurable.hasAutomation,
            configurable.editableIcon,
            configurable.taggable
        )
    }

    private fun getConfigurableType(configurable: Configurable): ConfigurableType {
        if (configurable is ConditionConfigurable) {
            return ConfigurableType.Condition
        }
        return if (configurable is StateDeviceConfigurable) {
            ConfigurableType.StateDevice
        } else ConfigurableType.Other
    }
}